### PR TITLE
Expose istio-agent metrics and remodel error handling

### PR DESCRIPTION
**Pull request by howardjohn**
Co-authored-by: "Aditya Prerepa" <adiprerepa@gmail.com>

Fixes https://github.com/istio/istio/issues/22825

This is a superset of #24798
#22318 (comment). If its
controversial we can split the two out, but there is a lot of
overlapping code between the two so I kept them together.

Basically, this adds istio agent metrics. Because of conflicts with
applications, we do some special things to export them with istio_agent_
prefix to avoid collisions.

Additionally, we stop returning errors if envoy or the app return errors
in the scrape. This avoids the situation where we suddenly drop envoy
metrics because the app is down or vis-versa, making the situation even
worse as we lose visibility. To add some extra visibility into this, we
also add metrics for total scrapes and failed scrapes.

### CODE REVIEW

File: pilot/cmd/pilot-agent/status/server.go [hunk 1]
```diff
@@ -243,30 +267,58 @@ type PrometheusScrapeConfiguration struct {
...
+	if _, err := w.Write(agent); err != nil {
+		log.Errorf("failed to write agent metrics: %v", err)
+		appScrapeErrors.Increment()
+	}
+}
+
+func scrapeAgentMetrics() ([]byte, error) {
+	buf := &bytes.Buffer{}
+	mfs, err := promRegistry.Gather()
+	enc := expfmt.NewEncoder(buf, expfmt.FmtText)
```
Comment (jacob-delgado) [inline pilot/cmd/pilot-agent/status/server.go:311]: From **https://godoc.org/github.com/prometheus/common/expfmt#NewEncoder** there should also be a call `defer enc.Close()` statement. It doesn't look like the interface implements Close although the documentation refers to it.

Comment (jacob-delgado) [reply pilot/cmd/pilot-agent/status/server.go:311]: Pulled down the branch. No such method exists. LGTM.

File: pilot/cmd/pilot-agent/status/server.go [hunk 2]
```diff
@@ -243,30 +267,58 @@ type PrometheusScrapeConfiguration struct {
...
-		return
+		envoyScrapeErrors.Increment()
 	}
 	if _, err := w.Write(application); err != nil {
 		log.Errorf("failed to write application metrics: %v", err)
+		appScrapeErrors.Increment()
+	}
+	if _, err := w.Write(agent); err != nil {
+		log.Errorf("failed to write agent metrics: %v", err)
+		appScrapeErrors.Increment()
```
Comment (douglas-reid) [inline pilot/cmd/pilot-agent/status/server.go:304]: should this be `agentScrapeErrors` ?

Comment (howardjohn) [reply pilot/cmd/pilot-agent/status/server.go:304]: Yes, thanks!

**Pull request merged**